drivers/tee/optee/optee.c

```diff
@@ -1126,23 +1127,94 @@ static int set_optee_method(const struct device *dev)
 	return 0;
 }
 
+static bool optee_check_uid(const struct device *dev)
+{
+	struct arm_smccc_res res;
+	struct optee_driver_data *data = (struct optee_driver_data *)dev->data;
+
+	data->smc_call(OPTEE_SMC_CALLS_UID, 0, 0, 0, 0, 0, 0, 0, &res);
+
+	if (res.a0 == OPTEE_MSG_UID_0 && res.a1 == OPTEE_MSG_UID_1 &&
+	    res.a2 == OPTEE_MSG_UID_2 && res.a3 == OPTEE_MSG_UID_3) {
+		return true;
+	}
+
+	return false;
+}
+
+static void optee_get_revision(const struct device *dev)
+{
+	struct optee_driver_data *data = (struct optee_driver_data *)dev->data;
+	union {
+		struct arm_smccc_res smccc;
+		struct optee_smc_call_get_os_revision_result result;
+	} res = { 0 };
+
+	data->smc_call(OPTEE_SMC_CALL_GET_OS_REVISION, 0, 0, 0, 0, 0, 0, 0,
+		  &res.smccc);
+
+	if (res.result.build_id) {
+		LOG_INF("OPTEE revision %lu.%lu (%08lx)", res.result.major,
+			res.result.minor, res.result.build_id);
+	} else {
+		LOG_INF("OPTEE revision %lu.%lu", res.result.major, res.result.minor);
+	}
+}
+
+static bool optee_exchange_caps(const struct device *dev, unsigned long *sec_caps)
+{
+	struct optee_driver_data *data = (struct optee_driver_data *)dev->data;
+	union {
+		struct arm_smccc_res smccc;
+		struct optee_smc_exchange_capabilities_result result;
+	} res;
+	unsigned long a1 = 0;
+
+	if (!IS_ENABLED(CONFIG_SMP) || arch_num_cpus() == 1) {
+		a1 |= OPTEE_SMC_NSEC_CAP_UNIPROCESSOR;
+	}
+
+	data->smc_call(OPTEE_SMC_EXCHANGE_CAPABILITIES, a1, 0, 0, 0, 0, 0, 0,
+		  &res.smccc);
+
+	if (res.result.status != OPTEE_SMC_RETURN_OK) {
+		return false;
+	}
+
+	*sec_caps = res.result.capabilities;
+	return true;
+}
+
 static int optee_init(const struct device *dev)
 {
 	struct optee_driver_data *data = dev->data;
 
 	if (set_optee_method(dev)) {
 		return -ENOTSUP;
 	}
-	/*
-	 * TODO At this stage driver should request and validate
-	 * capabilities from OP-TEE OS.
-	 */
 
 	sys_dlist_init(&data->notif);
 	k_mutex_init(&data->supp.mutex);
 	k_sem_init(&data->supp.reqs_c, 0, 1);
 	sys_dlist_init(&data->supp.reqs);
 
+	if (!optee_check_uid(dev)) {
+		LOG_ERR("OPTEE API UID mismatch");
+		return -EINVAL;
+	}
+
+	optee_get_revision(dev);
+
+	if (!optee_exchange_caps(dev, &data->sec_caps)) {
+		LOG_ERR("OPTEE capabilities exchange failed\n");
+		return -EINVAL;
+	}
+
+	if (!(data->sec_caps & OPTEE_SMC_SEC_CAP_DYNAMIC_SHM)) {
+		LOG_ERR("OPTEE does not support dynamic shared memory");
```

Comment:
-> static shared memory pool I suppose.

Reply:
I think the message is right here:
```
/*
 * Secure world supports commands "register/unregister shared memory",
 * secure world accepts command buffers located in any parts of non-secure RAM
 */
#define [OPTEE_SMC_SEC_CAP_DYNAMIC_SHM](https://elixir.bootlin.com/op-tee/latest/C/ident/OPTEE_SMC_SEC_CAP_DYNAMIC_SHM)		[BIT](https://elixir.bootlin.com/op-tee/latest/C/ident/BIT)(2)
```

and usage of this bit on OP-TEE OS side:
```
#if defined(CFG_CORE_DYN_SHM)
	dyn_shm_en = [core_mmu_nsec_ddr_is_defined](https://elixir.bootlin.com/op-tee/latest/C/ident/core_mmu_nsec_ddr_is_defined)();
	if (dyn_shm_en)
		[args](https://elixir.bootlin.com/op-tee/latest/C/ident/args)->[a1](https://elixir.bootlin.com/op-tee/latest/C/ident/a1) |= [OPTEE_SMC_SEC_CAP_DYNAMIC_SHM](https://elixir.bootlin.com/op-tee/latest/C/ident/OPTEE_SMC_SEC_CAP_DYNAMIC_SHM);
#endif
	[IMSG](https://elixir.bootlin.com/op-tee/latest/C/ident/IMSG)("Dynamic shared memory is %sabled", dyn_shm_en ? "en" : "dis");
```